python/pyspark/sql/chunk_api.py

```diff
@@ -0,0 +1,126 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+from collections import namedtuple
+
+import pyarrow as pa
+
+from pyspark.rdd import _create_local_socket
+from pyspark.sql import DataFrame
+from pyspark.sql import SparkSession
+from pyspark.serializers import read_with_length, write_with_length
+from pyspark.sql.pandas.serializers import ArrowStreamSerializer
+from pyspark.errors import PySparkRuntimeError
+
+
+ChunkMeta = namedtuple("ChunkMeta", ["id", "row_count", "byte_count"])
+
+
+def persist_dataframe_as_chunks(
+    dataframe: DataFrame, max_records_per_batch: int
+) -> list[ChunkMeta]:
+    """
+    Persist and materialize the spark dataframe as chunks, each chunk is an arrow batch.
+    It tries to persist data to spark worker memory firstly, if memory is not sufficient,
+    then it fallbacks to persist spilled data to spark worker local disk.
+    Return the list of tuple (chunk_id, chunk_row_count, chunk_byte_count).
+    This function is only available when it is called from spark driver process.
+    """
+    spark = dataframe.sparkSession
+    if spark is None:
+        raise PySparkRuntimeError("Active spark session is required.")
+
+    sc = spark.sparkContext
+    if sc.getConf().get("spark.python.dataFrameChunkRead.enabled", "false").lower() != "true":
```

Comment:
Is this SQL API or core API? If it's a SQL API, we should put this into `StaticSQLConf`

Reply:
Putting it in StaticSQLConf sounds good.

Reply:
if https://github.com/apache/spark/pull/44294#discussion_r1440018036 is adopted, we don't need the config.